### PR TITLE
🏃 Remove raw klog calls in controllers

### LINKED ISSUE #1203: Remove raw `klog` calls in controllers
/kind feature

**Describe the solution you'd like**
A reconciler should own its own logger and not depend on global functions to log. This will allow us to print log lines more consistently at the expensive of one extra parameter somewhere.

The logging should be pluggable. If I decide to write my own main.go and import the machine and controller managers I should be able to use whatever logger I bring so long as it satisfies the recommended logger interface.

We can use `github.com/go-logr/logr` for the logging interface since that seems to be the standard for cluster-api across the board. This issue should keep the logs as exact as possible (log levels, context, etc).

I would recommend this post in controller-runtime for anyone interested in logging and why `logr` may not be the best choice. https://github.com/kubernetes-sigs/controller-runtime/blob/master/TMP-LOGGING.md 

Out of scope: Any calls to klog outside of the controllers can be saved for another issue.

Note: Getting klogr to work is tricky. You will have to manually parse the `-v` flag and set the value on the logger. It is not automatic. Please test your changes by running the manager in a cluster and viewing logging output.

/good-first-issue
/priority important-longterm

<details><summary>My recommended workflow, entirely optional, can completely ignore this part</summary>

I would do this in two commits and one PR

1. Add a logging interface to the reconcilers but don't use them yet. In main.go we should use `klogr` to satisfy the interface created/used in the reconciler. (likely `go-logr/logr`).

2. After the infrastructure is set up, replace all of the `klog` calls with calls to the reconciler's logger. Ensure the controllers do not import `k8s.io/klog`.
</details>

### PR DESCRIPTION
**Pull request by tahsinrahman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1203

---
This change is [Reviewable](https://reviewable.io/reviews/kubernetes-sigs/cluster-api/1456)

### CODE REVIEW

**Comment by tahsinrahman** [comment]
@chuckha can you review if I did it right? I can do the same for other controllers then :)

File: controllers/cluster_controller.go [hunk 1]
```diff
@@ -84,7 +83,6 @@ func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager, options controlle
 
 func (r *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
 	ctx := context.Background()
-	_ = r.Log.WithValues("cluster", req.NamespacedName)
```
Comment (chuckha) [inline controllers/cluster_controller.go]: I'd recommend defining the log here and using this instead of `r.Log` in the function. This gives every log some nice context

**Comment by tahsinrahman** [comment]
> Is there a log getting passed into the reconcilers in `main.go`? That's the only other thing to check, otherwise looks great to me!!

yes, it is defined in main.go
https://github.com/kubernetes-sigs/cluster-api/blob/c6f4c59ca29a943b7543c072180822e1e9d127d0/main.go#L115

File: controllers/cluster_controller.go (hunk 1 above)
Comment (tahsinrahman) [reply controllers/cluster_controller.go]: then we need to pass `logger` to every functions that log things :confused: 
another way to do is what capa did, a scope object is defined with a field named `logger`, and it is passed down to every funtion

Comment (chuckha) [reply controllers/cluster_controller.go]: sorry, I meant for just this function.

Define the logger up front with the desired context (if applicable) and use it throughout the function. However, functions not defined on the reconciler *should* have a logger passed into them if they need one. Logging is a dependency and should be treated as such.

Comment (tahsinrahman) [reply controllers/cluster_controller.go]: oh, I got it :) thanks!

Comment (tahsinrahman) [reply controllers/cluster_controller.go]: can you do a quick review again?

Comment (chuckha) [reply controllers/cluster_controller.go]: this is looking great!

File: controllers/cluster_controller.go [hunk 2]
```diff
@@ -161,21 +161,23 @@ func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cl
...
 	children, err := r.listChildren(ctx, cluster)
 	if err != nil {
-		klog.Errorf("Failed to list children of cluster %s/%s: %v", cluster.Namespace, cluster.Name, err)
+		logger.Error(err, "Failed to list children")
 		return reconcile.Result{}, err
 	}
 
 	if len(children) > 0 {
-		klog.Infof("Cluster %s/%s still has %d children - deleting them first", cluster.Namespace, cluster.Name, len(children))
+		logger.Info("Cluster still has children - deleting them first", "no-of-childs", len(children))
```
Comment (ncdc) [inline controllers/cluster_controller.go]:
```suggestion
		logger.Info("Cluster still has children - deleting them first", "count", len(children))
```

File: controllers/cluster_controller.go [hunk 3]
```diff
@@ -186,10 +188,9 @@ func (r *ClusterReconciler) reconcileDelete(ctx context.Context, cluster *cluste
 
 			gvk := child.GetObjectKind().GroupVersionKind().String()
 
-			klog.Infof("Cluster %s/%s: deleting child %s %s", cluster.Namespace, cluster.Name, gvk, accessor.GetName())
+			logger.Info("deleting child", "gvk", gvk, "accessor-name", accessor.GetName())
```
Comment (ncdc) [inline controllers/cluster_controller.go]:
```suggestion
			logger.Info("deleting child", "gvk", gvk, "name", accessor.GetName())
```

File: controllers/cluster_controller.go [hunk 4]
```diff
@@ -186,10 +188,9 @@ func (r *ClusterReconciler) reconcileDelete(ctx context.Context, cluster *cluste
...
 			if err := r.Client.Delete(context.Background(), child); err != nil {
-				err = errors.Wrapf(err, "error deleting cluster %s/%s: failed to delete %s %s", cluster.Namespace, cluster.Name, gvk, accessor.GetName())
-				klog.Errorf(err.Error())
+				logger.Error(err, "error deleting accessor", "gvk", gvk, "name", accessor.GetName())
```
Comment (ncdc) [inline controllers/cluster_controller.go]:
```suggestion
				logger.Error(err, "error deleting resource", "gvk", gvk, "name", accessor.GetName())
```

File: controllers/cluster_controller.go [hunk 5]
```diff
@@ -256,7 +259,7 @@ func (r *ClusterReconciler) listChildren(ctx context.Context, cluster *clusterv1
 	eachFunc := func(o runtime.Object) error {
 		acc, err := meta.Accessor(o)
 		if err != nil {
-			klog.Errorf("Cluster %s/%s: couldn't create accessor for type %T: %v", cluster.Namespace, cluster.Name, o, err)
+			logger.Error(err, "couldn't create accessor", "accessor-type", fmt.Sprintf("%T", o))
```
Comment (ncdc) [inline controllers/cluster_controller.go]:
```suggestion
			logger.Error(err, "couldn't create accessor", "type", fmt.Sprintf("%T", o))
```

File: controllers/cluster_controller.go [hunk 6]
```diff
@@ -321,9 +326,11 @@ func (r *ClusterReconciler) reconcileControlPlaneInitialized(ctx context.Context
 // controlPlaneMachineToCluster is a handler.ToRequestsFunc to be used to enqueue requests for reconciliation
 // for Cluster to update its status.controlPlaneInitialized field
 func (r *ClusterReconciler) controlPlaneMachineToCluster(o handler.MapObject) []ctrl.Request {
+	logger := r.Log
```
Comment (ncdc) [inline controllers/cluster_controller.go]: I'd undo the changes in this function unless you're planning on defining `WithValues` somewhere

File: controllers/cluster_controller_phases.go [hunk 7]
```diff
@@ -100,7 +101,7 @@ func (r *ClusterReconciler) reconcileExternal(ctx context.Context, cluster *clus
 	// Add watcher for external object, if there isn't one already.
 	_, loaded := r.externalWatchers.LoadOrStore(obj.GroupVersionKind().String(), struct{}{})
 	if !loaded && r.controller != nil {
-		klog.Infof("Adding watcher on external object %q", obj.GroupVersionKind())
+		logger.Info("Adding watcher on external object", "object", obj.GroupVersionKind())
```
Comment (ncdc) [inline controllers/cluster_controller_phases.go]:
```suggestion
		logger.Info("Adding watcher on external object", "gvk", obj.GroupVersionKind())
```

**Comment by vincepri** [comment]
@tahsinrahman Can we open a new PR for `Remove unused function `isScalingEvent``?

File: controllers/machine_controller.go [hunk 8]
```diff
@@ -173,18 +177,23 @@ func (r *MachineReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cl
...
+	logger := r.Log.WithValues("machine", m.Name, "namespace", m.Namespace)
+	if cluster != nil {
+		logger = logger.WithValues("cluster", cluster.Name)
+	}
+
 	if err := r.isDeleteNodeAllowed(ctx, m); err != nil {
 		switch err {
 		case errNilNodeRef:
-			klog.V(2).Infof("Deleting node is not allowed for machine %q: %v", m.Name, err)
+			logger.Error(err, "Deleting node is not allowed")
```
Comment (chuckha) [inline controllers/machine_controller.go]: outside the scope of this PR but these should probably return err

Comment (ncdc) [reply controllers/machine_controller.go]: Returning err means it gets rereconciled with backoff. Not sure that's what we want?

Comment (chuckha) [reply controllers/machine_controller.go]: ah ok, if the check fails we can try to check again, otherwise it's fine to return nil and not reconcile again 👍

File: controllers/machineset_controller.go [hunk 9]
```diff
@@ -375,9 +377,9 @@ func (r *MachineSetReconciler) getNewMachine(machineSet *clusterv1.MachineSet) *
 }
 
 // shouldExcludeMachine returns true if the machine should be filtered out, false otherwise.
-func shouldExcludeMachine(machineSet *clusterv1.MachineSet, machine *clusterv1.Machine) bool {
+func shouldExcludeMachine(machineSet *clusterv1.MachineSet, machine *clusterv1.Machine, logger logr.Logger) bool {
 	if metav1.GetControllerOf(machine) != nil && !metav1.IsControlledBy(machine, machineSet) {
-		klog.V(4).Infof("%s not controlled by %v", machine.Name, machineSet.Name)
+		logger.V(4).Info("machine is not controlled by machineset", "machine", machine.Name)
```
Comment (wfernandes) [inline controllers/machineset_controller.go]: I noticed that in a majority of the places `logger.Info` messages are capitalized. If it is intentional that logging messages are capitalized whereas the error messages follow the [Go style convention](https://github.com/golang/go/wiki/CodeReviewComments#error-strings), this should be `Machine is...`.

File: controllers/machineset_controller.go [hunk 10]
```diff
@@ -445,7 +447,7 @@ func (r *MachineSetReconciler) MachineToMachineSets(o handler.MapObject) []ctrl.
 
 	m, ok := o.Object.(*clusterv1.Machine)
 	if !ok {
-		klog.Errorf("expected a Machine but got a %T", o.Object)
+		r.Log.Error(errors.New("did not get machine"), "got", fmt.Sprintf("%T", o.Object))
```
Comment (wfernandes) [inline controllers/machineset_controller.go]: We can probably leave the error as `nil` and just provide the message `expected to receive a machine object type` and the key value can be `ObjectType`.
See [logging errors](https://github.com/kubernetes-sigs/controller-runtime/blob/aac8bc7a3cbc56b99cc250d1506a0a39765b573d/TMP-LOGGING.md#logging-errors).

**Comment by vincepri** [comment]
What's the status of this work? I'd like to get it merged at some point before v0.3.0

**Comment by tahsinrahman** [comment]
> What's the status of this work? I'd like to get it merged at some point before v0.3.0

I replaced all `klog` calls with reconciler.Log. I tried kept log messages as it was before. Should i change it to lowercase as @wfernandes suggested?

**Comment by ncdc** [comment]
@tahsinrahman do you mean uppercase?

**Comment by tahsinrahman** [comment]
yes, the klog messages was uppercase before and I left them as they were, only replaced klog with r.Log

**Comment by wfernandes** [comment]
@tahsinrahman I was suggesting that the `Info` logs would be uppercased (which majority of them are except for the one I pointed out).
The `Error` logs should follow the Go style convention and remain lowercase. 
Hope that clarifies my comment.

**Comment by ncdc** [comment]
There's a subtle distinction between `error`s and error logs. The text of an `error` should be lowercase. The text of a log, be it info or error, should start with a capital letter. Please don't change error logs to lowercase 😄.

**Comment by tahsinrahman** [comment]
@ncdc thanks for your suggestions! I've updated the pr :)

File: controllers/machineset_controller.go (hunk 10 above)
Comment (tahsinrahman) [reply controllers/machineset_controller.go]: thanks! I've updated it with nil error, ptal :)

**Comment by chuckha** [comment]
This looks good to me, I went over the changes GitHub said I hadn't seen and didn't see anything bad. Maybe we can get this merged and iterate if we find any more issues with it?

**Comment by ncdc** [comment]
@tahsinrahman can you rebase and we'll get this in? Thanks!

**Comment by tahsinrahman** [comment]
@ncdc rebased!

**Comment by vincepri** [comment]
@tahsinrahman Can you squash down to a single commit? They seem all related
